File: scripts/check_mcp_config.js
#!/usr/bin/env node
/**
 * Lightweight checker for project MCP configuration (.cursor/mcp.json).
 * Does not print secret values.
 */

const fs = require("fs");
const path = require("path");

const REPO_ROOT = path.resolve(__dirname, "..");
const MCP_PATH = path.join(REPO_ROOT, ".cursor", "mcp.json");

const REQUIRED_SERVERS = [
  "chrome-devtools",
  "context7",
  "filesystem",
  "github",
  "playwright",
];

const DANGEROUS_PATH_PATTERNS = [
  /^\/$/,
  /^\\?$/,
  /^[A-Za-z]:[/\\]?$/,
  /^\$\{userHome\}$/,
  /^\/Users\/[^/]+$/,
  /^\/home\/[^/]+$/,
  /^C:\\Users\\[^\\]+$/i,
  /^\/Volumes$/,
  /^~$/,
];

const SECRET_VALUE_PATTERNS = [
  /ghp_[A-Za-z0-9_]+/,
  /github_pat_[A-Za-z0-9_]+/,
  /sk-[A-Za-z0-9]{20,}/,
  /^glpat-[A-Za-z0-9_-]+$/,
];

function collectStrings(obj) {
  if (typeof obj === "string") return [obj];
  if (Array.isArray(obj)) return obj.flatMap(collectStrings);
  if (obj && typeof obj === "object") {
    return Object.entries(obj).flatMap(([k, v]) => [k, ...collectStrings(v)]);
  }
  return [];
}

function isDangerousPath(value) {
  const normalized = String(value).trim();
  return DANGEROUS_PATH_PATTERNS.some((p) => p.test(normalized));
}

function looksLikeHardcodedSecret(value) {
  if (typeof value !== "string") return false;
  if (value.includes("${env:") || value.startsWith("${")) return false;
  return SECRET_VALUE_PATTERNS.some((p) => p.test(value));
}

function main() {
  const errors = [];
  const warnings = [];
  const summary = [];

  if (!fs.existsSync(MCP_PATH)) {
    errors.push(`Missing MCP config: ${path.relative(REPO_ROOT, MCP_PATH)}`);
    printReport(errors, warnings, summary);
    process.exit(1);
  }

  let data;
  try {
    data = JSON.parse(fs.readFileSync(MCP_PATH, "utf8"));
  } catch (err) {
    errors.push(`Invalid JSON in mcp.json: ${err.message}`);
    printReport(errors, warnings, summary);
    process.exit(1);
  }

  const servers = data.mcpServers;
  if (!servers || typeof servers !== "object" || Array.isArray(servers)) {
    errors.push("mcpServers must be an object");
    printReport(errors, warnings, summary);
    process.exit(1);
  }

  const names = Object.keys(servers).sort();
  summary.push(`servers (${names.length}): ${names.join(", ") || "(none)"}`);

  for (const required of REQUIRED_SERVERS) {
    if (!(required in servers)) {
      errors.push(`missing required server: ${required}`);
    }
  }

  for (const [name, cfg] of Object.entries(servers)) {
    if (!cfg || typeof cfg !== "object" || Array.isArray(cfg)) {
      errors.push(`server '${name}' must be an object`);
      continue;
    }
    const command = cfg.command ?? "?";
    const args = Array.isArray(cfg.args) ? cfg.args : [];
    let argPreview = args.slice(0, 4).map(String).join(" ");
    if (args.length > 4) argPreview += " ...";
    summary.push(`  - ${name}: command=${JSON.stringify(command)} args=[${argPreview}]`);

    const env = cfg.env;
    if (env && typeof env === "object") {
      for (const [key, val] of Object.entries(env)) {
        if (typeof val === "string" && looksLikeHardcodedSecret(val)) {
          errors.push(
            `server '${name}' env.${key} looks like a hardcoded secret; use \${env:...}`
          );
        }
      }
    }

    if (name === "filesystem" && Array.isArray(args)) {
      for (const arg of args) {
        if (typeof arg !== "string") continue;
        if (isDangerousPath(arg)) {
          errors.push(
            `filesystem MCP arg ${JSON.stringify(arg)} is dangerously broad; use \${workspaceFolder} only`
          );
        }
        if (["/", "\\", "C:\\", "C:/"].includes(arg)) {
          errors.push(`filesystem MCP must not grant ${JSON.stringify(arg)}`);
        }
      }
    }
  }

  for (const s of collectStrings(data)) {
    if (looksLikeHardcodedSecret(s)) {
      errors.push("config contains a value that looks like a hardcoded secret");
      break;
    }
  }

  const githubCfg = servers.github;
  if (githubCfg && typeof githubCfg === "object") {
    const env = githubCfg.env;
    const tokenRef =
      env && typeof env === "object"
        ? env.GITHUB_PERSONAL_ACCESS_TOKEN
        : undefined;
    if (!tokenRef) {
      warnings.push(
        "github MCP has no GITHUB_PERSONAL_ACCESS_TOKEN env mapping; set GITHUB_TOKEN locally for remote API access"
      );
    } else if (
      typeof tokenRef === "string" &&
      !tokenRef.includes("${env:")
    ) {
      warnings.push(
        "github MCP token should use ${env:GITHUB_TOKEN}, not a literal value"
      );
    }
  }

  printReport(errors, warnings, summary);
  process.exit(errors.length > 0 ? 1 : 0);
}

function printReport(errors, warnings, summary) {
  console.log("MCP config check");
  console.log(`  path: ${path.relative(REPO_ROOT, MCP_PATH)}`);
  console.log("  summary:");
  for (const line of summary) {
    console.log(`    ${line}`);
  }
  if (warnings.length) {
    console.log("  warnings:");
    for (const w of warnings) {
      console.log(`    - ${w}`);
    }
  }
  if (errors.length) {
    console.log("  errors:");
    for (const e of errors) {
      console.log(`    - ${e}`);
    }
  }
  console.log(`  result: ${errors.length ? "FAIL" : "PASS"}`);
}

main();

File: scripts/check_mcp_config.py
#!/usr/bin/env python3
"""Lightweight checker for project MCP configuration (.cursor/mcp.json)."""

from __future__ import annotations

import json
import re
import sys
from pathlib import Path

REPO_ROOT = Path(__file__).resolve().parent.parent
MCP_PATH = REPO_ROOT / ".cursor" / "mcp.json"

# Paths that must not be the sole filesystem grant.
DANGEROUS_PATH_PATTERNS = (
    re.compile(r"^/$"),
    re.compile(r"^\\?$"),  # empty
    re.compile(r"^[A-Za-z]:[/\\]?$"),
    re.compile(r"^\$\{userHome\}$"),
    re.compile(r"^/Users/[^/]+$"),
    re.compile(r"^/home/[^/]+$"),
    re.compile(r"^C:\\Users\\[^\\]+$", re.IGNORECASE),
)

SECRET_VALUE_PATTERNS = (
    re.compile(r"ghp_[A-Za-z0-9_]+"),
    re.compile(r"github_pat_[A-Za-z0-9_]+"),
    re.compile(r"sk-[A-Za-z0-9]{20,}"),
    re.compile(r"^glpat-[A-Za-z0-9_-]+$"),
)


def _collect_strings(obj: object) -> list[str]:
    if isinstance(obj, str):
        return [obj]
    if isinstance(obj, dict):
        out: list[str] = []
        for k, v in obj.items():
            out.append(str(k))
            out.extend(_collect_strings(v))
        return out
    if isinstance(obj, list):
        out: list[str] = []
        for item in obj:
            out.extend(_collect_strings(item))
        return out
    return []


def _is_dangerous_path(value: str) -> bool:
    normalized = value.strip()
    return any(p.match(normalized) for p in DANGEROUS_PATH_PATTERNS)


def _looks_like_hardcoded_secret(value: str) -> bool:
    if "${env:" in value or value.startswith("${"):
        return False
    return any(p.search(value) for p in SECRET_VALUE_PATTERNS)


def main() -> int:
    errors: list[str] = []
    warnings: list[str] = []
    summary: list[str] = []

    if not MCP_PATH.is_file():
        errors.append(f"Missing MCP config: {MCP_PATH.relative_to(REPO_ROOT)}")
        _print_report(errors, warnings, summary)
        return 1

    try:
        raw = MCP_PATH.read_text(encoding="utf-8")
        data = json.loads(raw)
    except json.JSONDecodeError as exc:
        errors.append(f"Invalid JSON in {MCP_PATH.name}: {exc}")
        _print_report(errors, warnings, summary)
        return 1

    servers = data.get("mcpServers")
    if not isinstance(servers, dict):
        errors.append("mcpServers must be an object")
        _print_report(errors, warnings, summary)
        return 1

    names = sorted(servers.keys())
    summary.append(f"servers ({len(names)}): {', '.join(names) if names else '(none)'}")

    if "playwright" not in servers:
        warnings.append("playwright server not configured (recommended for UI rounds)")

    for name, cfg in servers.items():
        if not isinstance(cfg, dict):
            errors.append(f"server '{name}' must be an object")
            continue
        command = cfg.get("command", "?")
        args = cfg.get("args", [])
        arg_preview = " ".join(str(a) for a in args[:4])
        if len(args) > 4:
            arg_preview += " ..."
        summary.append(f"  - {name}: command={command!r} args=[{arg_preview}]")

        env = cfg.get("env", {})
        if isinstance(env, dict):
            for key, val in env.items():
                if isinstance(val, str) and _looks_like_hardcoded_secret(val):
                    errors.append(
                        f"server '{name}' env.{key} looks like a hardcoded secret; use ${{env:...}}"
                    )

        if name == "filesystem" and isinstance(args, list):
            for arg in args:
                if not isinstance(arg, str):
                    continue
                if _is_dangerous_path(arg):
                    errors.append(
                        f"filesystem MCP arg {arg!r} is dangerously broad; "
                        "use ${workspaceFolder} only"
                    )
                if arg in ("/", "\\", "C:\\", "C:/"):
                    errors.append(f"filesystem MCP must not grant {arg!r}")

        if name == "chrome-devtools" and isinstance(args, list):
            blob = " ".join(str(a) for a in args)
            if (
                "chrome_devtools_mcp_light_novel" not in blob
                and "light_novel-chrome-profile" not in blob
                and "--userDataDir" not in blob
            ):
                warnings.append(
                    "chrome-devtools uses shared default profile risk; "
                    "prefer scripts/chrome_devtools_mcp_light_novel.sh "
                    "(see docs/mcp_isolation_strategy_light_novel.md)"
                )

    # Scan all string values for accidental secrets (without printing values).
    for s in _collect_strings(data):
        if _looks_like_hardcoded_secret(s):
            errors.append("config contains a value that looks like a hardcoded secret")
            break

    github_cfg = servers.get("github", {})
    if isinstance(github_cfg, dict):
        env = github_cfg.get("env", {})
        token_ref = env.get("GITHUB_PERSONAL_ACCESS_TOKEN", "") if isinstance(env, dict) else ""
        if not token_ref:
            warnings.append(
                "github MCP has no GITHUB_PERSONAL_ACCESS_TOKEN env mapping; "
                "set GITHUB_TOKEN locally for remote API access"
            )
        elif isinstance(token_ref, str) and "${env:" not in token_ref:
            warnings.append(
                "github MCP token should use ${env:GITHUB_TOKEN}, not a literal value"
            )

    _print_report(errors, warnings, summary)
    return 1 if errors else 0


def _print_report(errors: list[str], warnings: list[str], summary: list[str]) -> None:
    print("MCP config check")
    print(f"  path: {MCP_PATH.relative_to(REPO_ROOT)}")
    print("  summary:")
    for line in summary:
        print(f"    {line}")
    if warnings:
        print("  warnings:")
        for w in warnings:
            print(f"    - {w}")
    if errors:
        print("  errors:")
        for e in errors:
            print(f"    - {e}")
    status = "FAIL" if errors else "PASS"
    print(f"  result: {status}")


if __name__ == "__main__":
    sys.exit(main())
